pkgs/development/r-modules/generate-r-packages.R

```diff
@@ -67,48 +69,36 @@ formatPackage <- function(name, version, sha256, depends, imports, linkingTo) {
                     )
     depends <- unlist(strsplit(depends, split=",", fixed=TRUE))
     depends <- lapply(depends, gsub, pattern="([^ \t\n(]+).*", replacement="\\1")
-    depends <- lapply(depends, gsub, pattern=".", replacement="_", fixed=TRUE)
     depends <- depends[depends %in% knownPackages]
     depends <- lapply(depends, escapeName)
     depends <- paste(depends)
-    depends <- paste(sort(unique(depends)), collapse=" ")
-    paste0("  ", attr, " = derive2 { name=\"", name, "\"; version=\"", version, "\"; sha256=\"", sha256, "\"; depends=[", depends, "]; };")
+    depends <- sort(unique(depends))
+    list(name=unbox(name), version=unbox(version), sha256=unbox(sha256), depends=depends)
 }
 
-clusterExport(cl, c("nixPrefetch","readFormatted", "mirrorUrl", "mirrorType", "knownPackages"))
-
+clusterExport(cl, c("escapeName", "nixPrefetch", "readFormatted", "mirrorUrl", "mirrorType", "knownPackages"))
 pkgs <- pkgs[order(Package)]
 
 write(paste("updating", mirrorType, "packages"), stderr())
 pkgs$sha256 <- parApply(cl, pkgs, 1, function(p) nixPrefetch(p[1], p[2]))
-nix <- apply(pkgs, 1, function(p) formatPackage(p[1], p[2], p[18], p[4], p[5], p[6]))
-write("done", stderr())
+
+escapedNames <- sapply(pkgs$Package, escapeName)
+pkgs <- lapply(pkgs$Package, function(p) with(pkgs[p], formatPackage(Package, Version, sha256, Depends, Imports, LinkingTo)))
+names(pkgs) <- escapedNames
 
 # Mark deleted packages as broken
-setkey(readFormatted, V2)
-markBroken <- function(name) {
-  str <- paste0(readFormatted[name], collapse='"')
-  if(sum(grep("broken = true;", str)))
-    return(str)
-  write(paste("marked", name, "as broken"), stderr())
-  gsub("};$", "broken = true; };", str)
-}
-broken <- lapply(setdiff(readFormatted[[2]], pkgs[[1]]), markBroken)
-
-cat("# This file is generated from generate-r-packages.R. DO NOT EDIT.\n")
-cat("# Execute the following command to update the file.\n")
-cat("#\n")
-cat(paste("# Rscript generate-r-packages.R", mirrorType, ">new && mv new", packagesFile))
-cat("\n\n")
-cat("{ self, derive }:\n")
-cat("let derive2 = derive ")
-if (mirrorType == "cran") { cat("{  }")
-} else if (mirrorType == "irkernel") { cat("{}")
-} else { cat("{ biocVersion = \"", biocVersion, "\"; }", sep="") }
-cat(";\n")
-cat("in with self; {\n")
-cat(paste(nix, collapse="\n"), "\n", sep="")
-cat(paste(broken, collapse="\n"), "\n", sep="")
-cat("}\n")
+brokenPkgs <- lapply(readFormatted[setdiff(names(readFormatted), names(pkgs))], function(p)
+  list(name=unbox(p$name),
+       version=unbox(p$version),
+       sha256=unbox(p$sha256),
+       depends=p$depends,
+       broken=unbox(T)))
+
+# TODO: discuss ordering
+pkgs <- pkgs[order(names(pkgs))]
+brokenPkgs <- brokenPkgs[order(names(brokenPkgs))]
```

Comment:
I don't know what the correct ordering should be.
Previously pkgs were ordered asciibetically, which is not the same as what sort() or order() does AFAICT.

Reply:
It's already ordered on line 80, on the unescaped names. I don't think the order should change in the ensuite so the ordering should be the same as before without these lines.

Reply:
Okay, so it turns out `sort` and `order` respect the system locale settings which is undesirable.

I have pushed a changed that sets the locale to `C`.

My conjecture is that `setkey` uses the `C` locale by default, that's why everything is sorted already like it is in the original script.
I'm not sure what `setkey` does, so I'm just guessing this part.  

However, `sort` is used for computing `depends`, so it used the system locale for sorting, which is probably not desirable.

Reply:
`setkey` does not change order, it builds an index for subsequent queries. The original script would have the same locale issue.

Reply:
Okay, so it turns out, that yes, the table ordering functions always use the "C" locale.
See:
https://rdatatable.gitlab.io/data.table/reference/setkey.html#details
https://rdatatable.gitlab.io/data.table/reference/setorder.html#details

Reply:
And I found that this expression: `pkgs <- pkgs[order(Package)]` is a noop
probably because the original table's ordering somehow overrides the result's ordering.

However if I change this over to `pkgs <- pkgs[order(Package)]` then it actually returns a table that has its rows ordered according to the current locale.

Reply:
I mentioned this earlier, it is already sorted previously on the unescaped names on line 83. These lines shouldn't be necessary since it was already sorted, and `brokenpkgs` comes from the JSON which should also be sorted.

Reply:
`brokenPkgs is not sorted by default, because it keeps the order of the elements of `prevPkgs`, which is just the previous non-broken and broken packages appended together, which were sorted separately.
This means that sorting `brokenPkgs` is absolutely needed.

Reply:
I guess it depends on the sorting you're trying to achive. If you sort brokenPkgs then it won't be consistent with past behaviour. It doesn't really make much difference, I'm fine with either way.